### DISCUSSION TITLE
Extras - ruff-lsp formatting vs black

### DISCUSSION THREAD

**Discussion by karimlevallois** [LazyVim/LazyVim#2810] [Q&A] [answered]
Hi guys! 

I have been using black in my other configs, but ideally would prefer to use ruff-lsp to do the formatting instead. However, I have no idea how to get that working or if it is even possible? Has anyone else set this up?

Also, unless I am missing something I cannot get black (extras) formatter to work either using <leader>cf or on save?

**Comment by akthe-at** [comment]
All you need to do is get rid of black if you have it installed, and simply add  under the servers = {} of lspconfig the following:
 `ruff_lsp = {},`

After your next boot, make sure Mason has installed ruff_lsp for you, if so. You should be able to get automatic formatting on save (via lsp fallback) You do not even need to modify the conform.nvim configuration. You should be able to also use the default <Leader>cf to format with ruff lsp as well. I myself just find myself using ctrl-s a lot while writing code to constantly save and format. As long as there are no major syntax errors, it will format instantly. If you want to modify the ruff settings at all you will need to setup a global or per project pyproject.toml file.

**Reply by karimlevallois** [reply]
Thanks for coming back. So if I’m using the ‘python extra’, it should just work for formatting out of the box as that uses ruff_lsp alongside pyright/basedpyright?

**Reply by dpetka2001** [reply]
After you installed the Python Extra and you open a python file, you can confirm with either `:ConformInfo` or `:LazyFormatInfo` what formatter is being used. If you only installed the Python LazyExtra and not the Black LazyExtra, then the default formatter should be `ruff-lsp`.

**Reply by karimlevallois** [reply]
It's still not formatting for some strange reason.

This is me adding in some random spacing and errors being shown.

<img width="1680" alt="Screenshot 2024-03-23 at 16 14 10" src="https://github.com/LazyVim/LazyVim/assets/80129273/56192dd4-40b5-4bba-bf00-cd263e8e486d">

and here is the confirmation that formatting is enabled in ruff-lap.

<img width="1680" alt="Screenshot 2024-03-23 at 16 14 20" src="https://github.com/LazyVim/LazyVim/assets/80129273/31e6938d-e1bc-47d1-b26f-4f0915969d43">

<img width="1680" alt="Screenshot 2024-03-23 at 16 14 58" src="https://github.com/LazyVim/LazyVim/assets/80129273/dedbf746-9579-41ee-bb58-847633c65f78">

Neither <leader>cf or ctrl-s is working, the file is saved in the latter, but not formatted.

**Reply by dpetka2001** [reply]
See @akthe-at 's answer. These are syntax errors. I don't know if maybe a linter would help with these. Don't use one.

**Comment by akthe-at** [comment]
Yes as long as you don't have black setup, I forgot that was the default with the LazyVim python extra!

**Comment by akthe-at** [comment, marked as answer]
It will not fix format if there are syntax errors, just formatting issues only.  I can't tell from my phone if that's what's going on here.

**Reply by karimlevallois** [reply]
Ah that will be it then, I was maybe a tad aggressive in my testing. Just taken a basic file and it's corrected as expected. Thank you again!
